src/rotation_generator.jl

```diff
@@ -0,0 +1,258 @@
+"""
+    abstract type RotationGenerator{N,T} <: StaticMatrix{N,N,T}
+
+An abstract type representing `N`-dimensional rotation generator. More abstractly, they represent
+skew-symmetric real `N`×`N` matrices.
+"""
+abstract type RotationGenerator{N,T} <: StaticMatrix{N,N,T} end
+
+Base.@pure StaticArrays.Size(::Type{RotationGenerator{N}}) where {N} = Size(N,N)
+Base.@pure StaticArrays.Size(::Type{RotationGenerator{N,T}}) where {N,T} = Size(N,N)
+Base.@pure StaticArrays.Size(::Type{R}) where {R<:RotationGenerator} = Size(supertype(R))
+Base.adjoint(r::RotationGenerator) = -r
+Base.transpose(r::RotationGenerator{N,T}) where {N,T<:Real} = -r
+
+# Generate identity-matrix with SMatrix
+# Note that ones(RotationGenerator3,dims...) is not Array{<:RotationGenerator} but Array{<:StaticMatrix{3,3}}
+Base.one(::RotationGenerator{N,T}) where {N,T} = one(SMatrix{N,N,T})
+Base.one(::Type{RotationGenerator}) = error("The dimension of rotation is not specified.")
+Base.one(::Type{<:RotationGenerator{N}}) where N = one(SMatrix{N,N})
+Base.one(::Type{<:RotationGenerator{N,T}}) where {N,T} = one(SMatrix{N,N,T})
+Base.ones(::Type{R}) where {R<:RotationGenerator} = ones(R, ()) # avoid StaticArray constructor
+Base.ones(::Type{R}, dims::Base.DimOrInd...) where {R<:RotationGenerator} = ones(typeof(one(R)),dims...)
+Base.ones(::Type{R}, dims::NTuple{N, Integer}) where {R<:RotationGenerator, N} = ones(typeof(one(R)),dims)
+
+# Generate zero rotation matrix
+Base.zero(r::RotationGenerator) = zero(typeof(r))
+
+# difference
+Base.:-(r1::RotationGenerator, r2::RotationGenerator) = r1 + (-r2)
+
+# `convert` goes through the constructors, similar to e.g. `Number`
+Base.convert(::Type{R}, rot::RotationGenerator{N}) where {N,R<:RotationGenerator{N}} = R(rot)
+
+# RotationGenerator matrices should be skew-symmetric. Only the operations we define,
+# like addition, will stay as RotationGenerators, otherwise users will get an
+# SMatrix{3,3} (e.g. rot1 * rot2 -> SMatrix)
+Base.@pure StaticArrays.similar_type(::Union{R,Type{R}}) where {R <: RotationGenerator} = SMatrix{size(R)..., eltype(R), prod(size(R))}
+Base.@pure StaticArrays.similar_type(::Union{R,Type{R}}, ::Type{T}) where {R <: RotationGenerator, T} = SMatrix{size(R)..., T, prod(size(R))}
+
+# Useful for converting arrays of rotations to another rotation eltype, for instance.
+# Only works because parameters of all the rotations are of a similar form
+# Would need to be more sophisticated if we have arbitrary dimensions, etc
+@inline function Base.promote_op(::Type{R1}, ::Type{R2}) where {R1 <: RotationGenerator, R2 <: RotationGenerator}
```

Comment:
Is this right? The operator is not defined. Did you mean `promote_rule`?

Reply:
Ah, I was just referring to [`src/core_types.jl`](https://github.com/JuliaGeometry/Rotations.jl/blob/v1.1.0/src/core_types.jl#L68).
https://github.com/JuliaGeometry/Rotations.jl/commit/0ccc87629e00df8973d8d21cfc1994e3dfde1eb1#diff-55fa94d9eb6d180ed431c8aae0ad29c7100a0ad88cbd35dfc81239882a614e8dR39

Reply:
I've tried renaming `promote_op` -> `promote_rule` and `isleaftype` -> `isconcretetype`, but it doesn't work well here.
These seems unnecessary, so I've dropped both `promote_op` definitions from `src/core_types.jl` and `src/rotation_generator.jl`.